SalesforceCommerceCloud/pwa-kit
# Upgrade to React 18

# Description
This PR upgraded React to version 18, and Chakra version 2 and any related library to be compatible with React 18. The list of upgraded package includes:
- React 18.2.0
- React-dom 18.2.0
- "@types/react": "^18.2.0",
- "@types/react-dom": "^18.2.1",
- "@testing-library/react": "^14.0.0",
- "@testing-library/user-event": "^14.4.3",
- "@chakra-ui/icons": "^2.0.19",
- "@chakra-ui/react": "^2.6.0",
- "@chakra-ui/skip-nav": "^2.0.15",
- "@chakra-ui/system": "^2.5.6",
- "react-hook-form": "^7.43.9",
- "framer-motion": "^4.1.17",


The following packages are removed as the peerDependencies are not <= React 17
- @testing-library/render-hooks
- @wojtekmaj/enzyme-adapter-react-17



# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- HydrateRoot API has changed. We cant pass callback function as we are doing. Instead,  we can pass a prop into the Component to have it called after it is created. See [here](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1166/files#diff-f4ee7217dbf638680d6780866b7e72f56790717a5c6cc3ed1c982794a180ae38R123)
- Fix hydration error by creating a utils function to use window.HYDRATING. See [here](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1166/files#diff-98ae5e3b8f23295a74097c39c936f2956b809122a50b8fb9d50d0b82e6399d68R201)
- React hook form v7
     - errors object has been moved into formState object. see [here](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1166/files#diff-dcabbbe144fa55ab3119333bd3f022cb2b56cdfa280e762741e26a9ef2d2c8bdR41)
     - render prop will return an object which contains field and fieldState. See [here](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1166/files#diff-dcbda8bf3fa68c67d77d5af339bade5c86df5d1df5c9454a67397190218a33f5R48)
- For unit tests changes: 
   - There is a bug in a dependency of a dependency, nswapi, which is a dependency of jest-environment-jsdom that is not compatible with Chakra 2. It will throw an error `':disabled):not([disabled]' is not a valid selector` on some of the tests that has `Modal`. Solution: we use npm 8, and add overrides to enforce the version of nwsapi package to 2.2.2, and we have to drop npm 7 support
   - UserEvent actions are now async action, it will always return promise , and we have to call `setup` before using it
```
import userEvent from '@testing-library/user-event
test('mytest' async () => {
     const user = userEvent.setup()
     render(<Component />)
     
     
    await  user.click(screen.getByText(/something/i)
})
```   


# How to Test-Drive This PR

-  `npm ci` at root
- Start up Retail App
- Smoke test the app, and make sure it behaves as usual

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
